### ISSUE TITLE
A mistake in PyTorch Docs for nn.RNN

### ISSUE THREAD

**Issue by shehper** [pytorch/pytorch#129446]
### 📚 The doc issue

The code snippet in PyTorch docs for [`nn.RNN`](https://pytorch.org/docs/stable/generated/torch.nn.RNN.html#torch.nn.RNN) seems to have a mistake. 

Inside the `forward` function, `h_t[layer]` is defined as 

```
h_t[layer] = torch.tanh(
                x[t] @ weight_ih[layer].T
                + bias_ih[layer]
                + h_t_minus_1[layer] @ weight_hh[layer].T
                + bias_hh[layer]
            )
```

For `layer != 0`, we should write `h_t[layer-1]` instead of `x[t]`. A multi-layer RNN acts on the hidden state of the previous layer when layer > 0. 

### Suggest a potential alternative/fix

We could change the code snippet to the following. 

```
# Efficient implementation equivalent to the following with bidirectional=False
def forward(x, h_0=None):
    if batch_first:
        x = x.transpose(0, 1)
    seq_len, batch_size, _ = x.size()
    if h_0 is None:
        h_0 = torch.zeros(num_layers, batch_size, hidden_size)
    h_t_minus_1 = h_0
    h_t = h_0
    output = []
    for t in range(seq_len):
        for layer in range(num_layers):
            ih_input = x[t] if layer == 0 else h_t[layer-1]
            h_t[layer] = torch.tanh(
                ih_input @ weight_ih[layer].T
                + bias_ih[layer]
                + h_t_minus_1[layer] @ weight_hh[layer].T
                + bias_hh[layer]
            )
        output.append(h_t[-1].clone())
        h_t_minus_1 = h_t
    output = torch.stack(output)
    if batch_first:
        output = output.transpose(0, 1)
    return output, h_t
```

**Comment by drtonyr** [comment]
I agree that there appears to be a mistake.  I found this thread through https://stackoverflow.com/questions/78805977/multilayer-rnn-from-scratch-having-input-dilemma-and-inplace-replace-error-durin which confirms that others think this is a mistake as well.

I would like to also note that the use of `h_0 = torch.zeros(num_layers, batch_size, hidden_size)` is not correct for this implementation.  Whilst a single Tensor can be used in nn.RNN, it can't be used in this example code as the `t` loop overwrites the information needed to compute the gradient.   If a list of  `torch.zeros(batch_size, hidden_size)` is used instead then assignment does not replace and the gradient can be computed.

**Comment by jafraustro** [comment]
The issue reported that, for multi-layer RNNs, the input to each layer after the first should be the hidden state from the previous layer at the same time step, i.e., h_t[layer-1], not x[t]. The previous documentation incorrectly used x[t] for all layers. This was fixed in https://github.com/pytorch/pytorch/pull/140575 and improve in https://github.com/pytorch/pytorch/pull/153620.

The current PyTorch documentation snippet for nn.RNN is:
```python
# Efficient implementation equivalent to the following with bidirectional=False
def forward(x, hx=None):
    if batch_first:
        x = x.transpose(0, 1)
    seq_len, batch_size, _ = x.size()
    if hx is None:
        hx = torch.zeros(num_layers, batch_size, hidden_size)
    h_t_minus_1 = hx
    h_t = hx
    output = []
    for t in range(seq_len):
        for layer in range(num_layers):
            h_t[layer] = torch.tanh(
                x[t] @ weight_ih[layer].T
                + bias_ih[layer]
                + h_t_minus_1[layer] @ weight_hh[layer].T
                + bias_hh[layer]
            )
        output.append(h_t[-1])
        h_t_minus_1 = h_t
    output = torch.stack(output)
    if batch_first:
        output = output.transpose(0, 1)
    return output, h_t
```

This clarifies that for layer > 0, the input is the previous layer’s hidden state, and if no initial hidden state (hx) is provided, it is initialized to zeros.
